onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlDFT.h

```diff
@@ -0,0 +1,702 @@
+
+#include "../External/D3DX12/d3dx12.h"
+
+#include <wrl/client.h>
+#include <wrl/implements.h>
+
+#include <sstream>
+
+using namespace Microsoft::WRL;
+
+namespace DFTHelpers {
+    // Divides and rounds up
+    inline uint32_t CeilDivide(uint32_t dividend, uint32_t divisor)
+    {
+        UINT64 temp = static_cast<UINT64>(dividend) + divisor - 1;
+        return static_cast<uint32_t>(temp / divisor);
+    }
+
+    // Gets the next number of elements to dispatch to the GPU within a loop handling a large
+    // total number of tensor elements and threads.
+    void GetNextDispatchSize(
+        uint32_t elementCount,
+        uint32_t elementsPerThread,
+        uint32_t numThreads,
+        _Out_ uint32_t& dispatch,
+        _Out_ uint32_t& pendingElementCount
+    )
+    {
+        // Max threads per workgroup is 2^10 (1024). Max dispatch per dimension is 2^16. Taken together, we can dispatch a maximum of
+        // 2^26 (268,435,456) threads along a single dimension. This should suffice for a majority of the workload. Therefore, even
+        // though it is possible to dispatch up to (2^16)^3 workgroups simultaneously, we stick to the simpler 1D dispatch alternative.
+        assert(numThreads <= D3D12_CS_THREAD_GROUP_MAX_THREADS_PER_GROUP);
+
+        const uint32_t maxThreadsPerDispatch = numThreads * D3D12_CS_DISPATCH_MAX_THREAD_GROUPS_PER_DIMENSION;
+
+        const uint32_t requiredThreadCount = CeilDivide(elementCount, elementsPerThread);
+
+        // Compute max dispatchable elements
+        const uint32_t availableThreadCount = std::min(requiredThreadCount, maxThreadsPerDispatch);
+
+        // Compute required thread group count
+        uint32_t workGroupCount1D = CeilDivide(availableThreadCount, numThreads);
+
+        // Compute min dispatch size
+        dispatch = workGroupCount1D;
+
+        // With the dispatch size computed, compute the dispatched element count
+        const uint32_t dispatchedElementCount = workGroupCount1D * numThreads * elementsPerThread;
+
+        // Update the pending element count
+        pendingElementCount = (dispatchedElementCount < elementCount) ? elementCount - dispatchedElementCount : 0;
+    }
+
+}
+
+class GpuDFTOperator : public WRL::Base<IMLOperatorKernel>
+{
+private:
+    ComPtr<ID3D12Device> m_device;
+    ComPtr<ID3D12RootSignature> m_rootSignature;
+    ComPtr<ID3D12PipelineState> m_pipelineState;
+
+    std::vector<uint32_t> inputDims_ = {};
+    std::vector<uint32_t> outputDims_ = {};
+    int64_t axis_;
+    bool is_onesided_;
+    bool inverse_;
+
+    uint32_t outputDataSize_ = 0;
+    uint32_t inputDataSize_ = 0;
+    uint32_t outputIdx_ = 0;
+    uint32_t numPasses_ = 0;
+
+    // Allocate temporary buffers if needed
+    struct ResourceDesc {
+        ComPtr<ID3D12Resource> Resource;
+        std::array<uint32_t, 4> Sizes;
+        std::array<uint32_t, 4> Strides;
+    };
+    std::vector<ResourceDesc> resourceLoopList_ = {};
+
+    struct LoopRange {
```

Comment:
> { [](http://example.com/codeflow?start=20&length=2)

[nit] New line, like line 106. Elsewhere too (consistent with other files in this folder).